File: legacy/legacy-mahabharata-ios/Mahabharata/Model/DataClasses/Visual/Animations/RotateAnim.swift
//
//  Comics.swift
//  Mahabharata
//
//  Created by Roman Developer on 11/28/17.
//  Copyright © 2017 Iron Water Studio. All rights reserved.
//

import UIKit

extension BinaryInteger {
	var degreesToRadians: CGFloat { return CGFloat(Int(self)) * .pi / 180 }
}

extension FloatingPoint {
	var degreesToRadians: Self { return self * .pi / 180 }
	var radiansToDegrees: Self { return self * 180 / .pi }
}

class RotateAnim: Anim {
	let angle: Double
	let pivotX: Double
	let pivotY: Double
	
	private enum CodingKeys: String, CodingKey {
		case angle
		case pivotX
		case pivotY
	}
	
	required init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)
		self.angle = (try? container.decode(Double.self, forKey: .angle)) ?? 0
		self.pivotX = (try? container.decode(Double.self, forKey: .pivotX)) ?? 0
		self.pivotY = (try? container.decode(Double.self, forKey: .pivotY)) ?? 0
		
		try super.init(from: decoder)
	}
	
	init(angle: Double, pivotX: Double, pivotY: Double) {
		self.angle = angle
		self.pivotX = pivotX
		self.pivotY = pivotY
		
		super.init()
	}
	
	override var description: String {
		return [
			super.description,
			"Angle: \(angle)",
			"PivotX: \(pivotX)",
			"PivotY: \(pivotY)"
			].joined(separator: ", ")
	}
	
	//MARK: - Overrides
	override func interpolate(endAnim: Anim, fraction: Double) -> Anim {
		let end = endAnim as! RotateAnim
		return RotateAnim(angle: LinearInterpolation.evaluate(fraction: fraction, startValue: self.angle, endValue: end.angle),
		                  pivotX: LinearInterpolation.evaluate(fraction: fraction, startValue: self.pivotX, endValue: end.pivotX),
		                  pivotY: LinearInterpolation.evaluate(fraction: fraction, startValue: self.pivotY, endValue: end.pivotY))
	}
	
	override func apply(to data: Layer.ViewData, width: Int, height: Int) {
		//Because we have system anchor point == (0.5, 0.5)
		let pivotX = self.pivotX - 0.5
		let pivotY = self.pivotY - 0.5
		
		data.matrix = CATransform3DTranslate(data.matrix, CGFloat(Double(width) * pivotX), CGFloat(Double(height) * pivotY), 0)
		data.matrix = CATransform3DRotate(data.matrix, CGFloat(self.angle.degreesToRadians), 0, 0, 1)
		data.matrix = CATransform3DTranslate(data.matrix, CGFloat(-Double(width) * pivotX), CGFloat(-Double(height) * pivotY), 0)
	}
}

File: legacy/legacy-mahabharata-ios/Mahabharata/Model/DataClasses/Visual/Animations/RotateAnim.java
package com.fulldome.mahabharata.model.visual.animation;

import android.support.annotation.FloatRange;

import com.fulldome.mahabharata.model.visual.Layer;

public class RotateAnim extends LayerAnim {
	private float angle;
	private float pivotX;
	private float pivotY;

	public RotateAnim() {
	}

	public RotateAnim(float angle, float pivotX, float pivotY) {
		this.angle = angle;
		this.pivotX = pivotX;
		this.pivotY = pivotY;
	}

	@FloatRange(from = 0f, to = 360f)
	public float getAngle() {
		return angle;
	}

	@FloatRange(from = 0f, to = 1f)
	public float getPivotX() {
		return pivotX;
	}

	@FloatRange(from = 0f, to = 1f)
	public float getPivotY() {
		return pivotY;
	}

	@Override
	protected RotateAnim interpolate(LayerAnim endAnim, float fraction) {
		RotateAnim end = (RotateAnim) endAnim;
		return new RotateAnim(Layer.FLOAT_EVALUATOR.evaluate(fraction, getAngle(), end.getAngle()),
				Layer.FLOAT_EVALUATOR.evaluate(fraction, getPivotX(), end.getPivotX()),
				Layer.FLOAT_EVALUATOR.evaluate(fraction, getPivotY(), end.getPivotY()));
	}

	@Override
	public void apply(Layer.ViewData data, int width, int height) {
		data.getMatrix().postTranslate(-width * pivotX, -height * pivotY);
		data.getMatrix().postRotate(angle);
		data.getMatrix().postTranslate(width * pivotX, height * pivotY);
	}
}
